src/types/weakref/callableproxy.rs

```diff
@@ -0,0 +1,1000 @@
+    /// # Examples
+    #[cfg_attr(
+        not(all(feature = "macros", not(all(Py_LIMITED_API, not(Py_3_9))))),
+        doc = "```rust,ignore"
+    )]
+    #[cfg_attr(
+        all(feature = "macros", not(all(Py_LIMITED_API, not(Py_3_9)))),
+        doc = "```rust"
+    )]
+    /// use pyo3::prelude::*;
+    /// use pyo3::types::PyWeakCallableProxy;
+    ///
+    /// #[pyclass(weakref)]
+    /// struct Foo { /* fields omitted */ }
+    ///
+    /// #[pymethods]
+    /// impl Foo {
+    ///     fn __call__(&self) -> &str {
+    ///         "This class is callable"
+    ///     }
+    /// }
+    ///
+    /// #[pyfunction]
+    /// fn callback(wref: Bound<'_, PyWeakCallableProxy>) -> PyResult<()> {
+    ///         let py = wref.py();
+    ///         assert!(wref.upgrade_as::<Foo>()?.is_none());
+    ///         py.run_bound("counter = 1", None, None)
+    /// }
+    ///
+    /// # fn main() -> PyResult<()> {
+    /// Python::with_gil(|py| {
+    ///     py.run_bound("counter = 0", None, None)?;
+    ///     assert_eq!(py.eval_bound("counter", None, None)?.extract::<u32>()?, 0);
+    ///     let foo = Bound::new(py, Foo{})?;
+    ///
+    ///     // This is fine.
+    ///     let weakref = PyWeakCallableProxy::new_bound_with(&foo, py.None())?;
+    ///     assert!(weakref.upgrade_as::<Foo>()?.is_some());
+    ///     assert!(
+    ///         // In normal situations where a direct `Bound<'py, Foo>` is required use `upgrade::<Foo>`
+    ///         weakref.upgrade()
+    ///             .map_or(false, |obj| obj.is(&foo))
+    ///     );
+    ///     assert_eq!(py.eval_bound("counter", None, None)?.extract::<u32>()?, 0);
+    ///
+    ///     let weakref2 = PyWeakCallableProxy::new_bound_with(&foo, wrap_pyfunction!(callback, py)?)?;
+    ///     assert!(!weakref.is(&weakref2)); // Not the same weakref
+    ///     assert!(weakref.eq(&weakref2)?);  // But Equal, since they point to the same object
+    ///
+    ///     assert_eq!(weakref.call0()?.to_string(), "This class is callable");
+    ///
+    ///     drop(foo);
+    ///
+    ///     assert!(weakref.upgrade_as::<Foo>()?.is_none());
+    ///     assert_eq!(py.eval_bound("counter", None, None)?.extract::<u32>()?, 1);
+    ///     Ok(())
+    /// })
+    /// # }
+    /// ```
+    /// # Panics
+    /// This function panics if the provided object is not callable.
+    #[track_caller]
+    pub fn new_bound_with<'py, C>(
+        object: &Bound<'py, PyAny>,
+        callback: C,
+    ) -> PyResult<Bound<'py, PyWeakCallableProxy>>
+    where
+        C: ToPyObject,
+    {
+        // TODO: Must track caller be used here?
+        #[track_caller]
+        fn inner<'py>(
+            object: &Bound<'py, PyAny>,
+            callback: Bound<'py, PyAny>,
+        ) -> PyResult<Bound<'py, PyWeakCallableProxy>> {
+            assert!(
+                object.is_callable(),
+                "An object to be referenced by a PyWeakCallableProxy should be callable. Use PyWeakProxy instead."
+            );
```

Comment:
I wonder, given this footgun here and in `PyWeakProxy` constructors, should we consider merging these types into just `PyWeakProxy`? I guess the downside is that we don't match the C API so well, but maybe it works out as more ergonomic? I'm open to hearing your thoughts on this one, there might be very good reasons to keep them separate.

Reply:
I agree that it is a real footgun. Merging these types wouldn't be that weird, since the Python-side API doesn't really differentiate between them. And the C API provides a check function which globs them together.

My suggestion may be the following:
- Rename `PyWeakProxy` to `PyWeakNonCallableProxy`
- Create a new `PyWeakProxy` which can hold an enum to abstract over `PyWeakNonCallableProxy` and `PyWeakCallableProxy`

Reply:
Looking at the type definitions, it seems that the check for both of these is `PyWeakref_CheckProxy`. I think this means that it gets hard to distinguish these two types from Rust. 

What use cases do you envision for the specific types from the enum? If we can't think of any right now, one option might be to slim this type down a bit and just implement `PyTypeCheck` instead of `PyTypeInfo`, i.e. more like `PyNone`.

Reply:
(and have one single `PyWeakProxy` covering both callable and non-callable types.)

Reply:
That would probably make, the most sense.
I also recently figured out that the weakref type objects are actually part of the internal/private API, so they cannot be expected to be consistent in ABI3 or between patch versions, but it has been quite stable as far as I can see.

So maybe it would be good to move `PyWeakRef` also to the non-typeobject implementation, at least for the ABI3 target. We could still use the Type object pointer if we are compiling for a specific version, since it appears fairly stable(, while it is not guaranteed).

This would also allow work for PyPy, since they do export the Check functions.

An option I could for see is that we would follow the check functions:
- `PyWeak...` (Maybe `PyWeakAnyRef`?): Any Weakreference type, using `PyWeakref_Check`
- `PyWeakRef`: `weakref.ReferenceType` using `PyWeakref_CheckRef` for the specific normal reference type
- `PyWeakProxy`: Any weak Proxy type (Callable or Not) using `PyWeakref_CheckProxy`.

This would align with the C API, and probably with the way these type will be used, since I can imagine that a function could take a weakreference, without caring what the exact reference type would be, since it does not matter (that much) when using it from the C API.
And the bonus point that we wouldn't have to exclude PyPy.

Reply:
Sorry for the delay here by me yet again. I agree with the suggestion to move them all to the non-typeobject implementation. In general I'm trying hard to make PyO3 not need to depend on CPython private APIs.

We can always expand the implementation later if CPython makes the APIs public.

As for naming, I guess the one that will be used most commonly is "any weak rerference type", so let's call that one `PyWeakref` maybe? We could then have `PyWeakrefReference` and  `PyWeakrefProxy` for the specific subtypes.

So:
- `PyWeakref`
- `PyWeakrefReference`
- `PyWeakrefProxy`

How does that sound to you?

Reply:
The naming makes sense.

However I'm having some trouble doing the type objectless implementation.
You mentioned `PyNone` and `PyNotImplemented` as an example implementation however those are more like  '_singleton_' objects.

There is probably some way to do it and I want to figure it out. But do you have a hint?

(I also will not have much free time in the coming 3 weeks, so it is going to take a bit longer.)

Reply:
I have done it for `PyWeakref` and `PyWeakrefProxy` and renamed `PyWeakRef` to `PyWeakrefReference`.

I still need to make the changes for `PyWeakrefReference`; however, I have 2 findings:
- `PyWeakrefReference` is quite a long name, is this very practical to use or might `PyWeakrefRef` be better. I like the longer name for the clarity, but it feels clunky due to the length.
- `PyWeakrefReference` (or Pythons `weakref.ref`) is subclassable (The only subclassable `weakref` type). However, this non-typepointer implementation would not allow for subclassing.  The use case of subclassing is to add extra data to the weakreference. [Python weakref documentation mentioning this](https://docs.python.org/3/library/weakref.html#weak-reference-objects)

Changing the implementation of `PyWeakrefReference` would mean no subclassing from PyO3 (with the cost of using an internal API, which has been available since Python 2.2 (and probably will remain available since they are used in the check macros)), but making the change would allow for alternative interpreter support.

What do you think?

Reply:
@davidhewitt, The solution I thought pushed now does the following:
Only `PyWeakrefReference` uses the unstable API, only if you are compiling against a specific CPython version.
This keeps it subclassable, while allowing support for the other interpreters and the Limited API.